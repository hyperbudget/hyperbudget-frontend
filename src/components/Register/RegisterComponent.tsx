import * as React from 'react';
import { connect } from 'react-redux';
import * as Actions from '../../store/actions/Actions';
import { Redirect } from 'react-router';
import { APIError } from '../../lib/APIError/APIError';
import { ErrorComponent } from '../Error/Error';
import { State } from '../../lib/State/State';

interface RegisterComponentProps {
    doRegister: (params: {
        username: string,
        firstname: string,
        lastname: string,
        password: string,
    }) => void;
    isLoggedIn: boolean,
    token: string;
    loginErrors: APIError[],
};

class RegisterComponent extends React.Component<RegisterComponentProps, {}> {
    userNameRef: React.RefObject<HTMLInputElement>;
    passwordRef: React.RefObject<HTMLInputElement>;
    nameRef: React.RefObject<HTMLInputElement>;
    lastNameRef: React.RefObject<HTMLInputElement>;

    componentDidUpdate(prevProps, prevState) {
    }

    disableButton() {
        return !(
            this.userNameRef.current &&
            this.passwordRef.current &&
            this.nameRef.current &&
            this.userNameRef.current.value &&
            this.passwordRef.current.value &&
            this.nameRef.current.value
        );
    }

    constructor(props) {
        super(props);

        this.userNameRef = React.createRef();
        this.passwordRef = React.createRef();
        this.nameRef = React.createRef();
        this.lastNameRef = React.createRef();
    }

    render() {
        return (
        this.props.token
        ?
        <>
            <Redirect to="/report" />
        </>
        :
        <div className="home">
            <div className='home-text'>
                <form className='loginForm'>
                    {
                        (this.props.loginErrors ? <ErrorComponent errors={this.props.loginErrors} /> : '')
                    }
                    <div className='form-group'>
                        <label htmlFor="email">
                            Email:
                        </label>
                        <input className='form-control' id="email" name="email" type="email" ref={this.userNameRef} />
                    </div>
                    <div className='form-group'>
                        <label htmlFor="firstname">
                            Firstname:
                        </label>
                        <input className='form-control' id="firstname" name="firstname" type="text" ref={this.nameRef} />
                    </div>
                    <div className='form-group'>
                        <label htmlFor="lastname">
                            Last:
                        </label>
                        <input className='form-control' id="last" name="lastname" type="text" ref={this.lastNameRef} />
                    </div>
                    <div className='form-group'>
                        <label htmlFor="password">
                            Login Password:
                        </label>
                        <input className='form-control' id="password" type="password" name="password" ref={this.passwordRef} />
                    </div>
                    <div className='form-group'>
                        <input className='form-control btn btn-primary' type='button'  onClick={() => this.props.doRegister({
                            username: this.userNameRef.current.value,
                            firstname: this.nameRef.current.value,
                            lastname: this.lastNameRef.current.value,
                            password: this.passwordRef.current.value
                        }) } value="Login" />
                    </div>
                </form>
            </div>
        </div>
        );
    }
}


const mapStateToProps = (state: State) => {
    console.log(state);
    return {
        isLoggedIn: state.user.isLoggedIn,
        token: state.user.token,
        loginErrors: state.user.loginErrors,
    }
};

const mapDispatchToProps = dispatch => {
    return {
        doRegister: (params) => {
            return dispatch(Actions.do_register(
                {
                    firstname: params.firstname,
                    lastname: params.lastname,
                    username: params.username,
                    password: params.password,
                }
            ));
        },
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(RegisterComponent);
