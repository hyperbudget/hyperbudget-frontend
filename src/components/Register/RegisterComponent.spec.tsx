import React from 'react';
import { RegisterComponent } from './RegisterComponent';
import renderer from 'react-test-renderer';
import * as enzyme from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';

import 'jest';

beforeAll(() => {
  enzyme.configure({ adapter: new Adapter() });
});

test('Register component renders', () => {
  let component = renderer.create(
    <RegisterComponent
      doRegister={() => {}}
      isLoggedIn={false}
      token={null}
      APIError={null}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Button calls doRegister', () => {
  const mockRegister = jest.fn();

  let component = enzyme.mount(
    <RegisterComponent
      doRegister={mockRegister}
      isLoggedIn={false}
      token={null}
      APIError={null}
    />
  );

  component.find('#registerButton').simulate('click');

  expect(mockRegister.mock.calls.length).toBe(1);
  expect(mockRegister.mock.calls[0]).toMatchSnapshot();

  component.find('#email').instance()['value'] = 'errietta@errietta.me';
  component.find('#firstname').instance()['value'] = 'Errietta';
  component.find('#last').instance()['value'] = 'Kostala';
  component.find('#password').instance()['value'] = 'supersekrit';

  component.find('#registerButton').simulate('click');

  expect(mockRegister.mock.calls.length).toBe(2);
  expect(mockRegister.mock.calls[1]).toMatchSnapshot();

});

