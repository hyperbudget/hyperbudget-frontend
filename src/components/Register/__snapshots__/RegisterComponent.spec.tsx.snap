// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button calls doRegister 1`] = `
Array [
  Object {
    "firstname": "",
    "lastname": "",
    "password": "",
    "username": "",
  },
]
`;

exports[`Button calls doRegister 2`] = `
Array [
  Object {
    "firstname": "Errietta",
    "lastname": "Kostala",
    "password": "supersekrit",
    "username": "errietta@errietta.me",
  },
]
`;

exports[`Register component renders 1`] = `
<div
  className="home"
>
  <div
    className="home-text"
  >
    <form
      className="loginForm"
    >
      
      <div
        className="form-group"
      >
        <label
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="form-control"
          id="email"
          name="email"
          type="email"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="firstname"
        >
          Firstname:
        </label>
        <input
          className="form-control"
          id="firstname"
          name="firstname"
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="lastname"
        >
          Last:
        </label>
        <input
          className="form-control"
          id="last"
          name="lastname"
          type="text"
        />
      </div>
      <div
        className="form-group"
      >
        <label
          htmlFor="password"
        >
          Login Password:
        </label>
        <input
          className="form-control"
          id="password"
          name="password"
          type="password"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="form-control btn btn-primary"
          id="registerButton"
          onClick={[Function]}
          type="button"
          value="Login"
        />
      </div>
    </form>
  </div>
</div>
`;
